### PR TITLE
LeakyRelu Gradient

### PR DESCRIPTION
**Pull request by baijumeswani**
This pull request introduces the LeakyRelu Gradient.

### CODE REVIEW

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 1]
```diff
@@ -2004,5 +2004,26 @@ IMPLEMENT_GRADIENT_BUILDER(GetReciprocalGradient) {
...
+
+  return {zero_constant_node,
+          alpha_constant_node,
+          NodeDef("Sign", {I(0)}, {IA("Sign_I0")}),
+          NodeDef("Max", {zero, IA("Sign_I0")}, {IA("Max_Sign_I0")}),
+          NodeDef("Min", {zero, IA("Sign_I0")}, {IA("Min_Sign_I0")}),
+          NodeDef("Mul", {alpha, IA("Min_Sign_I0")}, {IA("Alpha_Min_Sign_I0")}),
+          NodeDef("Sub", {IA("Max_Sign_I0"), IA("Alpha_Min_Sign_I0")}, {IA("LeakyRelu_Grad")}),
+          NodeDef("Mul", {GO(0), IA("LeakyRelu_Grad")}, {GI(0)})};
+}
```
Comment (baijumeswani) [inline orttraining/orttraining/core/graph/gradient_builder.cc:2026]: Alternatively, I can use `Less->Where`. Let me know if there is a preference.

Comment (baijumeswani) [reply orttraining/orttraining/core/graph/gradient_builder.cc]: Went ahead and did this.

**Review by er3x3 — COMMENTED** [review]
I am thinking this one worths a contrib training Op for better perf as there are 3 Ops working on same size as input. PyTorch also has its own kernel for both CPU and GPU. (They use y to compare to 0 instead of using x, so the alpha must be a positive number?)

https://github.com/pytorch/pytorch/blob/3f4fee735a88d5ccc79b03acc4cc06c80940b90e/aten/src/ATen/native/cpu/Activation.cpp
https://github.com/pytorch/pytorch/blob/8f1c3c68d3aba5c8898bfb3144988aab6776d549/aten/src/ATen/native/cuda/ActivationLeakyReluKernel.cu

File: orttraining/orttraining/training_ops/cpu/op_gradients.cc [hunk 2]
```diff
@@ -260,5 +260,39 @@ Status QuickGeluGrad<T>::Compute(OpKernelContext* context) const {
...
+  concurrency::ThreadPool::TryBatchParallelFor(
+      tp, static_cast<int32_t>(task_count),
+      [&](ptrdiff_t task_idx) {
+        const auto start = task_idx * length_per_task;
+        const T* p_dy = dY_data + start;
+        const T* p_y = Y_data + start;
+        T* p_dx = dX_data + start;
+        int64_t count = std::min(length_per_task, elem_count - start);
+        for (int64_t i = 0; i < count; i++) {
+          p_dx[i] = p_dy[i] * (p_y[i] > 0.0f ? 1.0f : alpha_);
```
Comment (baijumeswani) [inline orttraining/orttraining/training_ops/cpu/op_gradients.cc:290]: Could not find a way to use `Eigen` `Map` with conditional

`dx = dy * (y > 0 ? 1.0f : alpha)`

Let me know if you are aware of any way to do that.

Comment (er3x3) [reply orttraining/orttraining/training_ops/cpu/op_gradients.cc:290]: Not sure if the sample code in onnxruntime/core/providers/cpu/activation/activation.h can help or not, something like, which use ".select()" function:

```
  void operator()(std::ptrdiff_t first, std::ptrdiff_t last) const final {
    ptrdiff_t len = last - first;
    T* output_ptr = this->output + first;
    ConstEigenVectorArrayMap<T> xm(this->input + first, len);
    EigenVectorArrayMap<T> ym(output_ptr, len);
    ym = (xm >= 0).select(xm, (T)alpha * (xm.exp() - 1));
  }
```

Comment (baijumeswani) [reply orttraining/orttraining/training_ops/cpu/op_gradients.cc]: Thanks. :)

**Review by er3x3 — APPROVED** [review]

**Comment by baijumeswani** [comment]
Thanks for the review @er3x3 ! 😀

**Comment by pauldog** [comment]
Hi what version of onnxruntime.training is LeakyRelu in? I can only get version 1.15.1 from pip install onnxruntime.training

**Pull request merged**